integration/build/build_test.go

```diff
@@ -0,0 +1,101 @@
+package build
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"io/ioutil"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/integration/util/request"
+)
+
+func TestBuildFailuresAndIntermediateContainers(t *testing.T) {
+
+	cases := []struct {
+		dockerfile                     string
+		numberOfIntermediateContainers int
+		rm                             bool
+		forceRm                        bool
+	}{
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 0`,
+			numberOfIntermediateContainers: 2,
+			rm:      false,
+			forceRm: false,
+		},
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 0`,
+			numberOfIntermediateContainers: 0,
+			rm:      true,
+			forceRm: false,
+		},
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 0`,
+			numberOfIntermediateContainers: 0,
+			rm:      true,
+			forceRm: true,
+		},
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 1`,
+			numberOfIntermediateContainers: 2,
+			rm:      false,
+			forceRm: false,
+		},
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 1`,
+			numberOfIntermediateContainers: 1,
+			rm:      true,
+			forceRm: false,
+		},
+		{
+			dockerfile: `FROM busybox
+			RUN exit 0
+			RUN exit 1`,
+			numberOfIntermediateContainers: 0,
+			rm:      true,
+			forceRm: true,
+		},
+	}
+
+	client := request.NewAPIClient(t)
+	ctx := context.Background()
+	for _, c := range cases {
+
+		existing, err := client.ContainerList(ctx, types.ContainerListOptions{All: true})
+		require.NoError(t, err)
+		dockerfile := []byte(c.dockerfile)
+
+		buff := bytes.NewBuffer(nil)
+		tw := tar.NewWriter(buff)
+		require.NoError(t, tw.WriteHeader(&tar.Header{
+			Name: "Dockerfile",
+			Size: int64(len(dockerfile)),
+		}))
+		_, err = tw.Write(dockerfile)
+		require.NoError(t, err)
+		require.NoError(t, tw.Close())
+		resp, err := client.ImageBuild(ctx, buff, types.ImageBuildOptions{Remove: c.rm, ForceRemove: c.forceRm, NoCache: true})
+		require.NoError(t, err)
+		ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+
+		new, err := client.ContainerList(ctx, types.ContainerListOptions{All: true})
+		require.NoError(t, err)
+		actualCountDiff := len(new) - len(existing)
+		require.Equal(t, c.numberOfIntermediateContainers, actualCountDiff, "Expected %v remaining intermediate containers, got %v. {rm: %v, force-rm: %v dockerfile:%v}", c.numberOfIntermediateContainers, actualCountDiff, c.rm, c.forceRm, c.dockerfile)
```

Comment:
If you use `t.Run()` you can use this extra string as the test name so that any assert/require failure has this extra information associated with it, instead of just the last assert.Equal()